test/pfm/pfm_test.go

```diff
@@ -0,0 +1,237 @@
+package pfm
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/celestiaorg/celestia-app/v2/app"
+	utils "github.com/celestiaorg/celestia-app/v2/test/tokenfilter"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/ibc-go/v6/modules/apps/transfer/types"
+	clienttypes "github.com/cosmos/ibc-go/v6/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v6/modules/core/04-channel/types"
+	host "github.com/cosmos/ibc-go/v6/modules/core/24-host"
+	ibctesting "github.com/cosmos/ibc-go/v6/testing"
+	"github.com/stretchr/testify/require"
+)
+
+type PacketMetadata struct {
+	Forward *ForwardMetadata `json:"forward"`
+}
+
+type ForwardMetadata struct {
+	Receiver       string        `json:"receiver"`
+	Port           string        `json:"port"`
+	Channel        string        `json:"channel"`
+	Timeout        time.Duration `json:"timeout"`
+	Retries        *uint8        `json:"retries,omitempty"`
+	Next           *string       `json:"next,omitempty"`
+	RefundSequence *uint64       `json:"refund_sequence,omitempty"`
+}
+
+func SetupTest(t *testing.T) (*ibctesting.Coordinator, *ibctesting.TestChain,
+	*ibctesting.TestChain, *ibctesting.TestChain,
+) {
+	chains := make(map[string]*ibctesting.TestChain)
+	coordinator := &ibctesting.Coordinator{
+		T:           t,
+		CurrentTime: time.Now(),
+		Chains:      chains,
+	}
+	celestiaChain := utils.NewTestChain(t, coordinator, ibctesting.GetChainID(1))
+	chainA := NewTestChain(t, coordinator, ibctesting.GetChainID(2))
+	chainB := NewTestChain(t, coordinator, ibctesting.GetChainID(3))
+	coordinator.Chains[ibctesting.GetChainID(1)] = celestiaChain
+	coordinator.Chains[ibctesting.GetChainID(2)] = chainA
+	coordinator.Chains[ibctesting.GetChainID(3)] = chainB
+	return coordinator, chainA, celestiaChain, chainB
+}
+
+func NewTransferPaths(chain1, chain2, chain3 *ibctesting.TestChain) (*ibctesting.Path, *ibctesting.Path) {
+	path1 := ibctesting.NewPath(chain1, chain2)
+	path1.EndpointA.ChannelConfig.PortID = ibctesting.TransferPort
+	path1.EndpointB.ChannelConfig.PortID = ibctesting.TransferPort
+	path1.EndpointA.ChannelConfig.Version = types.Version
+	path1.EndpointB.ChannelConfig.Version = types.Version
+	path2 := ibctesting.NewPath(chain2, chain3)
+	path2.EndpointA.ChannelConfig.PortID = ibctesting.TransferPort
+	path2.EndpointB.ChannelConfig.PortID = ibctesting.TransferPort
+	path2.EndpointA.ChannelConfig.Version = types.Version
+	path2.EndpointB.ChannelConfig.Version = types.Version
+
+	return path1, path2
+}
+
+// TestPacketForwardMiddlewareTransfer sends a PFM transfer originating from Celestia to ChainA, then back to Celestia and finally to ChainB.
+// It verifies that Celestia forwards the packet successfully, the balance of the sender account on Celestia decreases by the amount sent,
+// and the balance of the receiver account on ChainB increases by the amount sent.
```

Comment:
[out of curiosity] why Celestia -> chain A -> Celestia -> chain B?

Since we're only testing PFM, why not: chain A -> Celestia -> Chain B?

Reply:
because if we were to have this path: chain A -> Celestia -> Chain B tokenfilter would reject the transaction because it would not be tia but rather ibc denom prefixed with channel and port ids. 

In the way i did it first, when we initiate a transfer on Celestia and route that through chain A back to Celestia for forwarding it will unwind and become the native denom again.